core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -263,16 +265,15 @@ class Partition(val topic: String,
    *
    * This function can be triggered when a replica's LEO has incremented
    */
-  def maybeExpandIsr(replicaId: Int) {
+  def maybeExpandIsr(replicaId: Int, logReadResult: LogReadResult) {
     val leaderHWIncremented = inWriteLock(leaderIsrUpdateLock) {
       // check if this replica needs to be added to the ISR
       leaderReplicaIfLocal() match {
         case Some(leaderReplica) =>
           val replica = getReplica(replicaId).get
-          val leaderHW = leaderReplica.highWatermark
           if(!inSyncReplicas.contains(replica) &&
              assignedReplicas.map(_.brokerId).contains(replicaId) &&
-                  replica.logEndOffset.offsetDiff(leaderHW) >= 0) {
+             logReadResult.info.fetchOffsetMetadata.messageOffset >= Math.max(replica.lastLeaderLogEndOffset, logReadResult.hw)) {
```

Comment:
This approach solves problem stated in the ticket when replication factor >= 3, but it seems not solving the problem of replication factor = 2. When replication factor is 2 and ISR only contains leader, the high watermark will essentially be the log end offset of the leader. If we compare the fetch starting offset with the max of `lastLeaderLogEndOffset` and the current high watermark, it may still always appear to be lagging behind if there are small frequent produce requests.

This might not be that bad, as long as the replica is fetching faster than the producing rate, it will eventually catch up. I think this patch is an improvement but not solving all the issues. Let me think a bit more on this to see if we can have a more thorough solution.